File: overlay/python/hydra-kernel/src/main/python/hydra/python/lib/system.py
"""Python implementations of hydra.lib.system primitives.

These are real system-interface primitives. In Python the Hydra type ``effect<t>`` is
transparent (effect<t> = t), so each primitive performs its effect eagerly. Fallible
primitives return an ``Either[SystemError, T]``: a failure to launch (or perform the call)
is returned as ``Left(error)``; success is returned as ``Right(value)``. Infallible
primitives (``get_environment``, ``get_environment_variable``, ``get_time``) return their
value directly. This mirrors the Haskell reference implementation in
Hydra.Haskell.Lib.System. For #498.
"""

from __future__ import annotations
import os
import subprocess
import time

from hydra.dsl.python import Either, Left, Right, Given, None_
from hydra.python.util import PersistentMap
import hydra.system
import hydra.error.system as system_error
import hydra.file
import hydra.time

Command = hydra.system.Command
ProcessResult = hydra.system.ProcessResult
StatusCode = hydra.system.StatusCode
EnvironmentVariable = hydra.system.EnvironmentVariable
FilePath = hydra.file.FilePath
Timespec = hydra.time.Timespec


def execute(command: "hydra.system.Command") -> Either["system_error.SystemError", "hydra.system.ProcessResult"]:
    """Run a program to completion and capture its result.

    A child that runs and exits non-zero is returned as Right(result) with that exit code; only a
    failure to launch is Left(error). No shell is invoked; the program is executed directly.
    """
    program = command.program.value
    args = list(command.arguments)
    # working_directory and environment are Hydra optionals: Given(x) or None_().
    cwd = command.working_directory.value.value if isinstance(command.working_directory, Given) else None
    env = None
    if isinstance(command.environment, Given):
        env = {k.value: v for k, v in command.environment.value.items()}
    try:
        completed = subprocess.run(
            [program, *args],
            cwd=cwd,
            env=env,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            check=False)
    except FileNotFoundError:
        return Left(system_error.SystemErrorCommandNotFound(command.program))
    except NotADirectoryError:
        return Left(system_error.SystemErrorInvalidWorkingDirectory(command.program))
    except PermissionError:
        return Left(system_error.SystemErrorPermissionDenied(command.program))
    except OSError as e:
        return Left(system_error.SystemErrorOther(e.strerror or str(e)))
    return Right(ProcessResult(
        exit_code=StatusCode(completed.returncode),
        stdout=completed.stdout,
        stderr=completed.stderr))


def exit(code: "hydra.system.StatusCode") -> None:
    """Terminate the current process with the given status code. Does not return."""
    import sys
    sys.exit(code.value)


def get_environment():
    """Get the full set of environment variables, as a map from variable name to value."""
    return PersistentMap.from_pairs((EnvironmentVariable(k), v) for k, v in os.environ.items())


def get_environment_variable(name: "hydra.system.EnvironmentVariable"):
    """Look up a single environment variable by name; none if it is not set.

    Returns a Hydra optional (Given(value) or None_()).
    """
    value = os.environ.get(name.value)
    return None_() if value is None else Given(value)


def get_time() -> "hydra.time.Timespec":
    """Get the current wall-clock time as a Timespec (seconds and nanoseconds since the Unix epoch)."""
    nanos_total = time.time_ns()
    return Timespec(seconds=nanos_total // 1_000_000_000, nanoseconds=nanos_total % 1_000_000_000)


def get_working_directory() -> Either["system_error.SystemError", "hydra.file.FilePath"]:
    """Get the current working directory as a FilePath."""
    try:
        return Right(FilePath(os.getcwd()))
    except OSError as e:
        return Left(system_error.SystemErrorOther(e.strerror or str(e)))

File: overlay/scala/hydra-kernel/src/main/scala/hydra/scala/lib/system.scala
package hydra.scala.lib

import hydra.error.system.SystemError
import hydra.system.{Command, ProcessResult}

import java.io.{File, IOException}

/**
 * Scala implementations of hydra.lib.system primitives (#498).
 *
 * The effect type is transparent in Scala (effect<t> = t), so these helpers perform their work
 * eagerly. Fallible primitives return scala.util.Either[SystemError, T]: a launch failure becomes
 * Left(error); success becomes Right(value). Infallible primitives return their value directly.
 * Mirrors the classify pattern of the Haskell and Java hosts.
 *
 * Wrapped string/int types (hydra.file.FilePath, hydra.system.EnvironmentVariable, hydra.system.StatusCode)
 * are rendered by the Scala coder as plain String / Int, so they are used directly. The Hydra `binary`
 * type maps to a base64-encoded String in the Scala host, so captured stdout/stderr are base64-encoded.
 */
object system:
  /**
   * Run a program to completion, capturing stdout, stderr (base64), and the exit code. A child that
   * runs and exits non-zero is Right(result); only a failure to launch is Left(error). No shell is used.
   */
  def execute(command: Command): Either[SystemError, ProcessResult] =
    val argv = new java.util.ArrayList[String]()
    argv.add(command.program)
    command.arguments.foreach(argv.add)
    val pb = new ProcessBuilder(argv)
    command.workingDirectory.foreach(d => pb.directory(new File(d)))
    command.environment.foreach { env =>
      val pe = pb.environment()
      pe.clear()
      env.foreach { case (k, v) => pe.put(k, v) }
    }
    try
      val process = pb.start()
      val out = readAll(process.getInputStream)
      val err = readAll(process.getErrorStream)
      val code = process.waitFor()
      Right(ProcessResult(
        code,
        java.util.Base64.getEncoder.encodeToString(out),
        java.util.Base64.getEncoder.encodeToString(err)))
    catch
      case e: IOException          => Left(classify(command.program, e))
      case _: InterruptedException =>
        Thread.currentThread().interrupt()
        Left(SystemError.interrupted)

  /** Terminate the current process immediately with the given status. Does not return. */
  def exit(code: Int): Unit =
    System.exit(code)

  /** Return the entire environment of the current process as a map from variable name to value. */
  def getEnvironment: Map[String, String] =
    import scala.jdk.CollectionConverters.*
    System.getenv().asScala.toMap

  /** Return the value of the named environment variable, or None if it is not set. */
  def getEnvironmentVariable(name: String): Option[String] =
    Option(System.getenv(name))

  /** Return the current wall-clock time as a Timespec (seconds and nanoseconds since the Unix epoch). */
  def getTime: hydra.time.Timespec =
    val now = java.time.Instant.now()
    hydra.time.Timespec(now.getEpochSecond, now.getNano.toLong)

  /** Return the current working directory as a FilePath (String). */
  def getWorkingDirectory: Either[SystemError, String] =
    try Right(System.getProperty("user.dir"))
    catch case e: Exception => Left(SystemError.other(message(e)))

  // ---- Helpers (not primitives) ----

  private def readAll(in: java.io.InputStream): Array[Byte] =
    val buffer = new java.io.ByteArrayOutputStream()
    val chunk = new Array[Byte](8192)
    var n = in.read(chunk)
    while n != -1 do
      buffer.write(chunk, 0, n)
      n = in.read(chunk)
    buffer.toByteArray

  /** Classify a process-launch IOException into a SystemError. */
  private def classify(program: String, e: IOException): SystemError =
    val msg = message(e)
    val lower = msg.toLowerCase
    if lower.contains("error=2") || lower.contains("no such file") then SystemError.commandNotFound(program)
    else if lower.contains("error=13") || lower.contains("permission denied") then SystemError.permissionDenied(program)
    else if lower.contains("error=20") || lower.contains("not a directory") then SystemError.invalidWorkingDirectory(program)
    else SystemError.other(msg)

  private def message(e: Exception): String =
    val msg = e.getMessage
    if msg == null then e.getClass.getSimpleName else msg
